File: main/lru-implementation/lru-implementation-2.cs
#define DEBUG

using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;

internal sealed class LRU<T> : IEnumerable<T> {
    internal LRU(int capacity)
    {
        if (capacity <= 0) {
            throw new ArgumentOutOfRangeException(
                message: "capacity must be (strictly) positive",
                paramName: nameof(capacity));
        }
        
        Capacity = capacity;
    }
    
    internal int Capacity { get; }
    
    internal int Count => _chain.Count;
    
    public override string ToString() => string.Join(" ", this);
    
    public IEnumerator<T> GetEnumerator() => _chain.GetEnumerator();
    
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    
    internal bool Bump(T key) // returns true iff a cache miss occurs
    {
        LinkedListNode<T> node;
        if (_table.TryGetValue(key, out node)) {
            MoveToHead(node);
            return false;
        }
        
        if (Count == Capacity) {
            node = _chain.Last;
            _table.Remove(node.Value);
            MoveToHead(node);
            node.Value = key;
        } else {
            Contract.Assert(Count < Capacity);
            node = _chain.AddFirst(key);
        }
        
        _table.Add(key, node);
        return true;
    }
    
    private void MoveToHead(LinkedListNode<T> node)
    {
        _chain.Remove(node);
        _chain.AddFirst(node);
    }

    private readonly Dictionary<T, LinkedListNode<T>> _table
            = new Dictionary<T, LinkedListNode<T>>();
    
    private readonly LinkedList<T> _chain = new LinkedList<T>();
}

internal static class Solution {
    private static void Main()
    {
        var ns = Array.ConvertAll(ReadTokens(), int.Parse);
        var cache = new LRU<int>(ns[1]);
        Console.WriteLine(ReadValues(ns[0]).Count(cache.Bump));
        Console.WriteLine(cache);
    }
    
    private static IEnumerable<int> ReadValues(int count)
    {
        while (count > 0) {
            foreach (var key in ReadTokens().Select(int.Parse)) {
                yield return key;
                if (--count == 0) break;
            }
        }
    }
    
    private static string[] ReadTokens()
        => Console.ReadLine().Split((char[])null,
                                    StringSplitOptions.RemoveEmptyEntries);
}

File: main/lru-implementation/lru-implementation-2.cpp
#include <cassert>
#include <iostream>
#include <iterator>
#include <list>
#include <unordered_map>

namespace {
    class LRU {
    public:
        using size_type = std::list<int>::size_type;
        using const_iterator = std::list<int>::const_iterator;
        using iterator = const_iterator;

        explicit LRU(size_type capacity);

        size_type size() const noexcept { return chain_.size(); }
        size_type capacity() const noexcept { return capacity_; }

        const_iterator cbegin() const noexcept { return std::cbegin(chain_); }
        const_iterator cend() const noexcept { return std::cend(chain_); }
        iterator begin() const noexcept { return cbegin(); }
        iterator end() const noexcept { return cend(); }

        bool bump(int key); // returns true iff a cache miss occurred

    private:
        void move_to_front(const const_iterator node_it) noexcept
        {
            chain_.splice(cbegin(), chain_, node_it);
        }

        std::unordered_map<int, std::list<int>::iterator> table_ {};
        std::list<int> chain_ {};
        const size_type capacity_;
    };

    LRU::LRU(const size_type capacity) : capacity_{capacity}
    {
        if (capacity == 0u)
            throw std::out_of_range{"capacity must be positive"};
    }

    bool LRU::bump(const int key)
    {
        const auto row_it = table_.find(key);

        if (row_it != std::cend(table_)) {
            move_to_front(row_it->second);
            return false;
        }

        if (size() == capacity()) {
            const auto node_it = std::prev(std::end(chain_));
            table_.erase(*node_it);
            *node_it = key;
            move_to_front(node_it);
            table_.emplace(key, node_it);
        } else {
            assert(size() < capacity());
            chain_.push_front(key);
            table_.emplace(key, std::begin(chain_));
        }

        return true;
    }

    std::ostream& operator<<(std::ostream& out, const LRU& cache)
    {
        auto sep = "";
        for (const auto& x : cache) {
            out << sep << x;
            sep = " ";
        }

        return out;
    }
}

int main()
{
    int n {};
    LRU::size_type s {};
    std::cin >> n >> s;

    LRU cache {s};
    auto faults = 0;

    while (n-- > 0) {
        int x {};
        std::cin >> x;
        if (cache.bump(x)) ++faults;
    }

    std::cout << faults << '\n' << cache << '\n';
}
